sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,758 @@
+
+        log_msg += "max depth = {}, ".format(max_depth)
+
+        if self.do_early_stopping_:
+            name = 'neg-loss' if self.scoring == 'loss' else 'score'
+            log_msg += "train {}: {:.5f}, ".format(name, self.train_score_[-1])
+            if self.validation_fraction is not None:
+                log_msg += "val {}: {:.5f}, ".format(
+                    name, self.validation_score_[-1])
+
+        iteration_time = time() - iteration_start_time
+        log_msg += "in {:0.3f}s".format(iteration_time)
+
+        print(log_msg)
+
+    def _raw_predict(self, X):
+        """Return the sum of the leaves values over all predictors.
+
+        Parameters
+        ----------
+        X : array-like, shape=(n_samples, n_features)
+            The input samples.
+
+        Returns
+        -------
+        raw_predictions : array, shape (n_samples * n_trees_per_iteration,)
+            The raw predicted values.
+        """
+        X = check_array(X, dtype=[X_DTYPE, X_BINNED_DTYPE])
+        check_is_fitted(self, '_predictors')
+        if X.shape[1] != self.n_features_:
+            raise ValueError(
+                'X has {} features but this estimator was trained with '
+                '{} features.'.format(X.shape[1], self.n_features_)
+            )
+        is_binned = getattr(self, '_in_fit', False)
+        n_samples = X.shape[0]
+        raw_predictions = np.zeros(
+            shape=(self._n_trees_per_iteration, n_samples),
+            dtype=self._baseline_prediction.dtype
+        )
+        raw_predictions += self._baseline_prediction
+        for predictors_of_ith_iteration in self._predictors:
+            for k, predictor in enumerate(predictors_of_ith_iteration):
+                predict = (predictor.predict_binned if is_binned
+                           else predictor.predict)
+                raw_predictions[k, :] += predict(X)
+
+        return raw_predictions
+
+    @abstractmethod
+    def _get_loss(self):
+        pass
+
+    @abstractmethod
+    def _encode_y(self, y=None):
+        pass
+
+    @property
+    def n_iter_(self):
+        check_is_fitted(self, '_predictors')
+        return len(self._predictors)
+
+
+class HistGradientBoostingRegressor(BaseHistGradientBoosting, RegressorMixin):
+    """Histogram-based Gradient Boosting Regression Tree.
+
+    This estimator is much faster than
+    :class:`GradientBoostingRegressor<sklearn.ensemble.GradientBoostingRegressor>`
+    for big datasets (n_samples >= 10 000). The input data ``X`` is pre-binned
+    into integer-valued bins, which considerably reduces the number of
+    splitting points to consider, and allows the algorithm to leverage
+    integer-based data structures. For small sample sizes,
+    :class:`GradientBoostingRegressor<sklearn.ensemble.GradientBoostingRegressor>`
+    might be preferred since binning may lead to split points that are too
+    approximate in this setting.
+
```

Comment:
Do we add a link to a user guide entry?

Reply:
There's not much in the user guide about these new estimators yet (I'm planning to do that in another PR)